src/utils/Atomic.hxx

```diff
@@ -95,12 +95,34 @@ public:
     /// Locks the specific critical section.
     void lock()
     {
-        portENTER_CRITICAL(&mux);
+        // This should really use portENTER_CRITICAL_SAFE() but that is not
+        // available prior to ESP-IDF 3.3 which is not available in the
+        // arduino-esp32 environment generally. The below code is the
+        // implementation of that macro.
+        if (xPortInIsrContext())
+        {
+            portENTER_CRITICAL_ISR(&mux);
+        }
+        else
+        {
+            portENTER_CRITICAL(&mux);
+        }
     }
     /// Unlocks the specific critical section.
     void unlock()
     {
-        portEXIT_CRITICAL(&mux);
+        // This should really use portEXIT_CRITICAL_SAFE() but that is not
+        // available prior to ESP-IDF 3.3 which is not available in the
+        // arduino-esp32 environment generally. The below code is the
+        // implementation of that macro.
+        if (xPortInIsrContext())
+        {
+            portEXIT_CRITICAL_ISR(&mux);
```

Comment:
if you check the implementation of this, you will see that it is exactly the same as the other branch.

Reply:
For 3.2 yes, but in IDF 3.3/4.0 it is [changing](https://github.com/espressif/esp-idf/blob/6517c5033aab27afe491e433d6beeba93dfb1aa7/components/freertos/include/freertos/portmacro.h#L265-L294). This is a bit of future proofing .

Reply:
If/When portENTER_CRITICAL_SAFE/portEXIT_CRITICAL_SAFE is available it would be better to move to that though.